rfd/0abc-kube-access-forward-identity.md

```diff
@@ -0,0 +1,181 @@
+
+## What
+
+This RFD proposes a way to forward user identity from the Teleport Proxy to
+Teleport Kubernetes Service and remote Teleport proxy.
+
+### Related issues
+
+- [#21609](https://github.com/gravitational/teleport/issues/21609)
+
+## Why
+
+When a user connects to a Teleport proxy, the proxy will authenticate the user
+using X.509 certificate. Kubernetes Proxy will then forward the request to the
+Teleport service responsible for serving the Kubernetes cluster (Kube Service) or
+to the remote Teleport proxy when the Kubernetes Cluster belongs to a trusted cluster.
+For the proxy to be able to forward the user's identity to the correct service,
+it needs to generate a new X.509 key-pair with the user identity embedded that
+will be used during the mTLS authentication.
+The upstream service will then use the user identity to authorize the request and to
+impersonate the configured Kubernetes principals when forwarding the request to
+the Kubernetes API server.
+
+To avoid having to generate a new certificate for each request, the proxy will
+generate a new certificate for each user's identity and cache it in memory for
+`min(3h, certExpiration)`. This certificate cache must be keyed with a key that
+uniquely identifies the user's identity so that the proxy can retrieve the certificate
+for the correct identity when forwarding the request. In the past, we had situations
+where the cache was not properly keyed and the proxy would forward the request
+with the wrong certificate ending up with authorization errors.
+
+Currently, the key used to cache the certificate is the user's identity has the
+following format:
+
+```go
+func (c *authContext) key() string {
+	// it is important that the context key contains user, kubernetes groups and certificate expiry,
+	// so that new logins with different parameters will not reuse this context
+	return fmt.Sprintf("%v:%v:%v:%v:%v:%v:%v", c.teleportCluster.name, c.User.GetName(), c.kubeUsers, c.kubeGroups, c.kubeClusterName, c.certExpires.Unix(), c.Identity.GetIdentity().ActiveRequests)
+}
+```
+
+Each time we introduce a new field to the user's identity, we need to make sure
+that the key is updated to include the new field if the field is used by Kubernetes
+Access. This process is error-prone and it's easy to forget to update the key, and
+eventually introduce a bug or a security vulnerability.
+
+This RFD proposes a way to forward the user's identity from the proxy to the
+Kube Service and remote Teleport proxy that does not require the proxy to generate
+a new certificate for each user's identity.
+
+## Details
+
+The Proxy will authenticate to the Kube Service and remote Teleport proxy using its
+certificate. The proxy certificate does not contain any user identification
+and it's signed by the Teleport CA. The proxy will then forward the user's identity
+using HTTP headers or Proxy Protocol extensions.
+
+This change will remove the requirement for the proxy to generate a new certificate
```

Comment:
Is the Kube Service going to sign the client certificate for the target Kubernetes Cluster or what client credentials will it use? I think including a some sort of ascii flow chart diagram of the authentication/impersonation flow (i.e. who's generating certs and adding impersonation headers, who's terminating connection, who's forwarding to who ,etc.) would be useful to see for the local and leaf cluster scenarios. For example, see a diagram Steve made in https://github.com/gravitational/teleport/pull/22347 - makes it easy to see the entire flow.

Reply:
Kube Service connection to the target cluster will be the same. Currently, the request is authorized using the SA token the user created in the cluster with Impersonation permissions.

The only thing that differs from user to user is the `Impersonate-User` and `Impersonate-Groups` header values we set in the request. But that does not block us from re-using the same connection.

I will include the diagrams tomorrow